### DISCUSSION TITLE
[Question]: Token usage of Dalle-3 has not been tracked？

### DISCUSSION THREAD

**Discussion by yanxiyue** [danny-avila/LibreChat#1479] [Q&A]
### What is your question?

The document states that since v6.0.0, LibreChat has accurately tracked the token usage of OpenAI/Plugins endpoints, but it seems that the token usage of Dalle-3 has not been tracked

### More Details

 I checked the code in DALLE3.js, didn't see the any implementation。

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Token usage of prompts, completions, and functions are as accurate as they can be. For example, when you make a DALL-E request, it creates an OpenAI function, and generates a completed function when it's done; both are counted towards usage in a separate module.

DALL-E is specific to OpenAI. While there is an API cost, these are not technically "tokens" in the context of string values. It's the same with vision (image attachments).

The balance system would benefit from counting these against the users' credits but that is not yet implemented.

**Reply by juicesharp** [reply]
What about the Google Search plugin? I opened a plugin and ran a simple query in a chat in such configuration: 
![2024-01-05_17-44-57](https://github.com/danny-avila/LibreChat/assets/7575510/c63160a3-b8de-425d-85e9-2a4720f4ab92)
I got response in the chat and my credits were not updated. Is that expected?

**Reply by danny-avila** [reply]
@juicesharp it should be. Can you share your debug logs to confirm? They are in /api/logs or you may need to inspect your container logs

**Reply by juicesharp** [reply]
The chat is deployed on our company private virtual server, I may ask our dev-ops look there by tomorrow. Anyway I cannot reproduce that anymore, might be a client side glitch.

**Comment by danny-avila** [comment]
I just wanted to add to the discussion that since this PR was merged, Vision requests are now counted towards balance usage correctly since they are technically counted as tokens, and can be counted accurately each time: https://github.com/danny-avila/LibreChat/pull/1535

**Comment by alanyukeroo** [comment]
Hi @danny-avila I saw your previous response regarding the token usage tracking for DALL-E requests and the mention that the system would benefit from counting these API costs against user credits. Could you please confirm if this has now been implemented? Is the usage of DALL-E currently counted towards the users' credits?

Thank you!
